File: verify-dev-injection/inject_check.py
"""Injection-check for sdk-python. See run.sh for context.

Boots a DEFAULT-config datadir client (NO enable_quonfig_user_context, NO
QUONFIG_DEV_CONTEXT) and asserts the dev-override flag resolves purely from
token-file injection.
"""

from __future__ import annotations

import os
import sys


def fail(msg: str) -> "None":
    print(f"FAIL sdk-python: {msg}", file=sys.stderr)
    sys.exit(1)


fixture = os.environ.get("QFG_INJECT_FIXTURE_HOME")
if not fixture:
    fail("QFG_INJECT_FIXTURE_HOME unset")

# sdk-python's loader honors QUONFIG_CONFIG_HOME (and HOME as a fallback);
# point both at the fixture so the synthetic tokens.json is discovered (or,
# in the no-token phase, not found).
os.environ["HOME"] = fixture
os.environ["QUONFIG_CONFIG_HOME"] = fixture
# The default must hold without the env opt-in.
os.environ.pop("QUONFIG_DEV_CONTEXT", None)

datadir = os.environ.get("QFG_INJECT_DATADIR")
key = os.environ.get("QFG_INJECT_KEY")
expected = os.environ.get("QFG_INJECT_EXPECTED") == "true"
if not (datadir and key):
    fail("missing QFG_INJECT_* env vars")

try:
    from quonfig import Quonfig
except Exception as e:  # noqa: BLE001
    fail(f"import quonfig failed: {type(e).__name__}: {e}")

try:
    # DEFAULT config — deliberately NO enable_quonfig_user_context.
    client = Quonfig(
        datadir=datadir,
        environment="Production",
        collect_evaluation_summaries=False,
        context_upload_mode="none",
    ).init()
except Exception as e:  # noqa: BLE001
    fail(f"Quonfig() construct/init raised: {type(e).__name__}: {e}")

try:
    value = client.get_bool(key, default=False)
except Exception as e:  # noqa: BLE001
    fail(f"get_bool raised: {type(e).__name__}: {e}")

if value != expected:
    phase = "token-present" if expected else "no-token"
    fail(f"get_bool({key!r}) = {value!r}, expected {expected} (phase: {phase}, HOME={fixture})")

phase = "token-present" if expected else "no-token"
print(f"OK sdk-python: {phase} -> get_bool({key!r})={value}")

File: verify-dev-injection/inject_check.mjs
// Injection-check for sdk-node. See run.sh for context.
// Boots a DEFAULT-config datadir client (NO enableQuonfigUserContext, NO
// QUONFIG_DEV_CONTEXT) and asserts the dev-override flag resolves purely
// from token-file injection.
import process from "node:process";
import { pathToFileURL } from "node:url";
import path from "node:path";

function fail(msg) {
  console.error(`FAIL sdk-node: ${msg}`);
  process.exit(1);
}

const fixture = process.env.QFG_INJECT_FIXTURE_HOME;
if (!fixture) fail("QFG_INJECT_FIXTURE_HOME unset");
// os.homedir() reads $HOME first on Unix; point it at the fixture so the
// loader's ~/.quonfig/tokens.json lookup hits our synthetic file (or, in
// the no-token phase, finds nothing) without disturbing the outer HOME.
process.env.HOME = fixture;
// Belt-and-suspenders: the default must hold without the env opt-in.
delete process.env.QUONFIG_DEV_CONTEXT;

const datadir = process.env.QFG_INJECT_DATADIR;
const key = process.env.QFG_INJECT_KEY;
const expected = process.env.QFG_INJECT_EXPECTED === "true";
if (!datadir || !key) fail("missing QFG_INJECT_* env vars");

const here = path.dirname(new URL(import.meta.url).pathname);
const sdkEntry = path.resolve(here, "..", "..", "sdk-node", "dist", "index.js");
let mod;
try {
  mod = await import(pathToFileURL(sdkEntry).href);
} catch (e) {
  fail(`import @quonfig/node from ${sdkEntry}: ${e?.stack || e}`);
}
const { Quonfig } = mod;
if (!Quonfig) fail("sdk-node did not export Quonfig");

let q;
try {
  // DEFAULT config — deliberately NO enableQuonfigUserContext.
  q = new Quonfig({
    datadir,
    environment: "Production",
    collectEvaluationSummaries: false,
    contextUploadMode: "none",
  });
  await q.init();
} catch (e) {
  fail(`construct/init raised: ${e?.stack || e}`);
}

let value;
try {
  value = q.getBool(key);
} catch (e) {
  fail(`q.getBool(${key}) raised: ${e?.stack || e}`);
}
if (value !== expected) {
  fail(
    `q.getBool(${JSON.stringify(key)}) = ${JSON.stringify(value)}, expected ${expected} ` +
      `(phase: ${expected ? "token-present" : "no-token"}, HOME=${fixture})`
  );
}

console.log(
  `OK sdk-node: ${expected ? "token-present" : "no-token"} -> getBool(${JSON.stringify(key)})=${value}`
);
